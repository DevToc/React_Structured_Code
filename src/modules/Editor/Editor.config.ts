export interface ZoomOption {
  label: string;
  value: number;
}

export const ZOOM_OPTIONS: ZoomOption[] = [
  { label: '25%', value: 0.25 },
  { label: '40%', value: 0.4 },
  { label: '60%', value: 0.6 },
  { label: '80%', value: 0.8 },
  { label: '100%', value: 1 },
  { label: '120%', value: 1.2 },
  { label: '140%', value: 1.4 },
];

export enum IgnoreEventElement {
  INPUT = 'INPUT',
  TEXTAREA = 'TEXTAREA',
  SELECT = 'SELECT',
}

export enum IgnoreEventRole {
  switch = 'switch',
  slider = 'slider',
  radio = 'radio',
  range = 'range',
  tab = 'tab',
}

export const PAGE_CONTAINER_CLASSNAME = 'page-container';
export const EDITOR_TOOLBAR_ID = 'editor-toolbar-menu';
export const EDITOR_ACTIVE_PAGE_FOCUS_ID = 'editor-active-page-focus-id';

// TODO: The current export needs to be optimized for the design with many pages
// Limit it to 50 based on the metric and until we optimize the export
export const MAXIMUM_PAGE_LIMIT = 50;
