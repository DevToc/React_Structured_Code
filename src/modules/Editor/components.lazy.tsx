import { lazy } from 'react';

const NavbarDesktop = lazy(() => import('./components/Navbar').then(({ Navbar }) => ({ default: Navbar })));
const NavbarMobile = lazy(() => import('./components/Navbar').then(({ NavbarMobile }) => ({ default: NavbarMobile })));
const PageAreaDesktop = lazy(() => import('./components/PageArea').then(({ PageArea }) => ({ default: PageArea })));
const PageAreaMobile = lazy(() =>
  import('./components/PageArea').then(({ PageAreaMobile }) => ({ default: PageAreaMobile })),
);
const WidgetMenuDesktop = lazy(() =>
  import('./components/WidgetMenu').then(({ WidgetMenu }) => ({ default: WidgetMenu })),
);
const WidgetMenuMobile = lazy(() =>
  import('./components/WidgetMenu').then(({ WidgetMenuMobile }) => ({ default: WidgetMenuMobile })),
);

export const ResponsiveComponentsMap = {
  mobile: {
    navbar: NavbarMobile,
    widget: WidgetMenuMobile,
    pageArea: PageAreaMobile,
  },
  desktop: {
    navbar: NavbarDesktop,
    widget: WidgetMenuDesktop,
    pageArea: PageAreaDesktop,
  },
};
