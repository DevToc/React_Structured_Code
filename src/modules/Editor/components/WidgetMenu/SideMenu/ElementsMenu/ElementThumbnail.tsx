import { Flex, Image } from '@chakra-ui/react';

import { STYLED_RESPONSIVE_TEXT_WIDGETS } from './ElementsMenu.config';

interface ElementThumbnailProps {
  index: number;
  onClick: (index: number) => void;
}

export const ElementThumbnail = ({ index, onClick }: ElementThumbnailProps) => {
  const styleData = STYLED_RESPONSIVE_TEXT_WIDGETS[index];

  const handleClick = () => {
    onClick?.(index);
  };

  if (!styleData) return null;

  const { thumbnail } = styleData;

  return (
    <Flex
      w={'full'}
      h={'full'}
      direction={'column'}
      gap={1}
      p={4}
      alignItems={'center'}
      justifyContent={'center'}
      as={'button'}
      border={'1px solid'}
      borderColor={'divider.gray'}
      borderRadius={'base'}
      onClick={handleClick}
      transitionDuration={'200ms'}
      _hover={{
        bg: 'hover.gray',
      }}
      data-testid={`elementsmenu-${index}-btn`}
    >
      <Image src={thumbnail.src} objectFit={'contain'} h={'full'} alt={thumbnail.altText} />
    </Flex>
  );
};
