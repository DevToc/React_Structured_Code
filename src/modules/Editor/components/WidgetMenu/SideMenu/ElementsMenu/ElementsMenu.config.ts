import SetYourGoalsThumbnail from './elementThumbnails/set_your_goals.png';
import MonthlyActiveUsersThumbnail from './elementThumbnails/monthly_active_users.png';
import DesignThumbnail from './elementThumbnails/design.png';
import ReferralTrafficThumbnail from './elementThumbnails/referral_traffic.png';
import AnalysisThumbnail from './elementThumbnails/analysis.png';
import MarketConditionsThumbnail from './elementThumbnails/market_conditions.png';
import AdditionalUnitsThumbnail from './elementThumbnails/additional_units.png';
import QuoteThumbnail from './elementThumbnails/quote.png';

import { BorderStyle } from 'widgets/BasicShapeWidget/BasicShapeWidget.types';

// TODO - this should probably be stored in the DB instead of hardcode
// Default data for the styled widgets in the Elements category
export const STYLED_RESPONSIVE_TEXT_WIDGETS = [
  // set your goals
  {
    widthPx: 322,
    heightPx: 176.77,
    backgroundShapeData: {
      fillColor: ['rgba(2, 152, 166, 0.13)'],
      border: {
        color: '#0298A6',
        width: 5,
        style: BorderStyle.Solid,
      },
    },
    textWidgetData: {
      widthPx: 282,
      heightPx: 137,
      proseMirrorData: {
        type: 'doc',
        content: [
          {
            type: 'paragraph',
            attrs: {
              lineHeight: '1.2',
              textAlign: 'left',
            },
            content: [
              {
                type: 'text',
                marks: [
                  {
                    type: 'bold',
                  },
                  {
                    type: 'textStyle',
                    attrs: {
                      fontFamily: 'Karla',
                      fontSize: '18px',
                      color: '#2B2B35',
                    },
                  },
                ],
                text: 'Set your goals',
              },
            ],
          },
          {
            type: 'title',
            attrs: {
              lineHeight: '1.2',
              textAlign: 'left',
            },
            content: [
              {
                type: 'text',
                marks: [
                  {
                    type: 'textStyle',
                    attrs: {
                      fontFamily: 'Karla',
                      fontSize: '14px',
                      color: '#2B2B35',
                    },
                  },
                ],
                text: "Setting goals is an important part of success. It's not just about having a vague idea that you want to be healthy, fit, or whatever. If you're going to achieve your goals, you need to make them specific and measurable.",
              },
            ],
          },
        ],
      },
    },
    fontsToLoad: ['Karla'],
    thumbnail: {
      src: SetYourGoalsThumbnail,
      altText: 'Set your goals thumbnail',
    },
  },
  // 1.3 million users
  {
    widthPx: 244,
    heightPx: 217.59,
    backgroundShapeData: {
      fillColor: ['#7DC6EA'],
      border: {
        color: '#fff',
        width: 5,
        style: BorderStyle.Solid,
      },
    },
    textWidgetData: {
      widthPx: 204,
      heightPx: 177.59,
      proseMirrorData: {
        type: 'doc',
        content: [
          {
            type: 'paragraph',
            attrs: {
              lineHeight: '1.2',
              textAlign: 'left',
            },
            content: [
              {
                type: 'text',
                marks: [
                  {
                    type: 'textStyle',
                    attrs: {
                      fontFamily: 'Courier New',
                      fontSize: '15px',
                      color: '#1A4C7A',
                    },
                  },
                ],
                text: 'Our company has attracted more than',
              },
            ],
          },
          {
            type: 'title',
            attrs: {
              lineHeight: '1.2',
              textAlign: 'left',
            },
            content: [
              {
                type: 'text',
                marks: [
                  {
                    type: 'bold',
                  },
                  {
                    type: 'textStyle',
                    attrs: {
                      fontFamily: 'Montserrat',
                      fontSize: '49px',
                      color: '#1A4C7A',
                    },
                  },
                ],
                text: '1.3 Million',
              },
            ],
          },
          {
            type: 'title',
            attrs: {
              lineHeight: '1.2',
              textAlign: 'left',
            },
            content: [
              {
                type: 'text',
                marks: [
                  {
                    type: 'textStyle',
                    attrs: {
                      fontFamily: 'Courier New',
                      fontSize: '15px',
                      color: '#1A4C7A',
                    },
                  },
                ],
                text: 'Monthly Active Users',
              },
            ],
          },
        ],
      },
    },
    fontsToLoad: ['Courier New', 'Montserrat'],
    thumbnail: {
      src: MonthlyActiveUsersThumbnail,
      altText: '1.3 Million users',
    },
  },
  // design button
  {
    widthPx: 157,
    heightPx: 64,
    backgroundShapeData: {
      fillColor: ['#003F7F'],
      cornerRadius: 10,
      border: {
        color: 'rgba(0,0,0,0)',
        width: 0,
        style: BorderStyle.Solid,
      },
    },
    textWidgetData: {
      widthPx: 117,
      heightPx: 24,
      proseMirrorData: {
        type: 'doc',
        content: [
          {
            type: 'paragraph',
            attrs: {
              lineHeight: '1.2',
              textAlign: 'center',
            },
            content: [
              {
                type: 'text',
                marks: [
                  {
                    type: 'bold',
                  },
                  {
                    type: 'textStyle',
                    attrs: {
                      fontFamily: 'Rubik',
                      fontSize: '20px',
                      color: '#fff',
                    },
                  },
                ],
                text: 'Design',
              },
            ],
          },
        ],
      },
    },
    fontsToLoad: ['Rubik'],
    thumbnail: {
      src: DesignThumbnail,
      altText: 'Design button',
    },
  },
  // +25.5%
  {
    widthPx: 189,
    heightPx: 108,
    backgroundShapeData: {
      fillColor: ['#4FC6FF'],
      cornerRadius: 10,
      border: {
        color: 'rgba(0,0,0,0)',
        width: 0,
        style: BorderStyle.Solid,
      },
    },
    textWidgetData: {
      widthPx: 149,
      heightPx: 68.39,
      proseMirrorData: {
        type: 'doc',
        content: [
          {
            type: 'paragraph',
            attrs: {
              lineHeight: '1.2',
              textAlign: 'center',
            },
            content: [
              {
                type: 'text',
                marks: [
                  {
                    type: 'textStyle',
                    attrs: {
                      fontFamily: 'Vidaloka',
                      fontSize: '40px',
                      color: '#2B2B35',
                    },
                  },
                ],
                text: '+25.5%',
              },
            ],
          },
          {
            type: 'title',
            attrs: {
              lineHeight: '1.2',
              textAlign: 'center',
            },
            content: [
              {
                type: 'text',
                marks: [
                  {
                    type: 'bold',
                  },
                  {
                    type: 'textStyle',
                    attrs: {
                      fontFamily: 'Inter',
                      fontSize: '17px',
                      color: '#2B2B35',
                    },
                  },
                ],
                text: 'Referral Traffic',
              },
            ],
          },
        ],
      },
    },
    fontsToLoad: ['Vidaloka'],
    thumbnail: {
      src: ReferralTrafficThumbnail,
      altText: '25.5% of people',
    },
  },
  // Analysis
  {
    widthPx: 350,
    heightPx: 179.13,
    backgroundShapeData: {
      fillColor: ['#FFC72D'],
      border: {
        color: '#fff',
        width: 5,
        style: BorderStyle.Solid,
      },
    },
    textWidgetData: {
      widthPx: 310,
      heightPx: 139.13,
      proseMirrorData: {
        type: 'doc',
        content: [
          {
            type: 'paragraph',
            attrs: {
              lineHeight: '1.2',
              textAlign: 'center',
            },
            content: [
              {
                type: 'text',
                marks: [
                  {
                    type: 'bold',
                  },
                  {
                    type: 'textStyle',
                    attrs: {
                      fontFamily: 'Maven Pro',
                      fontSize: '20px',
                      color: '#2B2B35',
                    },
                  },
                ],
                text: 'Analysis',
              },
            ],
          },
          {
            type: 'paragraph',
            attrs: {
              lineHeight: '1.2',
              textAlign: 'center',
            },
          },
          {
            type: 'bulletList',
            content: [
              {
                type: 'listItem',
                content: [
                  {
                    type: 'paragraph',
                    attrs: {
                      lineHeight: '1.2',
                      textAlign: 'left',
                    },
                    content: [
                      {
                        type: 'text',
                        marks: [
                          {
                            type: 'textStyle',
                            attrs: {
                              fontFamily: 'Maven Pro',
                              fontSize: '15px',
                              color: '#2B2B35',
                            },
                          },
                        ],
                        text: 'Enhance employee’s performance, productivity and creativity',
                      },
                    ],
                  },
                ],
              },
            ],
          },
          {
            type: 'bulletList',
            content: [
              {
                type: 'listItem',
                content: [
                  {
                    type: 'paragraph',
                    attrs: {
                      lineHeight: '1.2',
                      textAlign: 'left',
                    },
                    content: [
                      {
                        type: 'text',
                        marks: [
                          {
                            type: 'textStyle',
                            attrs: {
                              fontFamily: 'Maven Pro',
                              fontSize: '15px',
                              color: '#2B2B35',
                            },
                          },
                        ],
                        text: 'Improve business outcomes',
                      },
                    ],
                  },
                ],
              },
              {
                type: 'listItem',
                content: [
                  {
                    type: 'paragraph',
                    attrs: {
                      lineHeight: '1.2',
                      textAlign: 'left',
                    },
                    content: [
                      {
                        type: 'text',
                        marks: [
                          {
                            type: 'textStyle',
                            attrs: {
                              fontFamily: 'Maven Pro',
                              fontSize: '15px',
                              color: '#2B2B35',
                            },
                          },
                        ],
                        text: 'Experience level: new team members',
                      },
                    ],
                  },
                ],
              },
              {
                type: 'listItem',
                content: [
                  {
                    type: 'paragraph',
                    attrs: {
                      lineHeight: '1.2',
                      textAlign: 'left',
                    },
                    content: [
                      {
                        type: 'text',
                        marks: [
                          {
                            type: 'textStyle',
                            attrs: {
                              fontFamily: 'Maven Pro',
                              fontSize: '15px',
                              color: '#2B2B35',
                            },
                          },
                        ],
                        text: 'Update inventory of existing content',
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
    },
    fontsToLoad: ['Maven Pro'],
    thumbnail: {
      src: AnalysisThumbnail,
      altText: 'Analysis',
    },
  },
  // Market Conditions
  {
    widthPx: 242,
    heightPx: 366.33,
    backgroundShapeData: {
      fillColor: ['#FAFFE4'],
      border: {
        color: '#000',
        width: 2,
        style: BorderStyle.Solid,
      },
    },
    textWidgetData: {
      widthPx: 202,
      heightPx: 326.33,
      proseMirrorData: {
        type: 'doc',
        content: [
          {
            type: 'paragraph',
            attrs: {
              lineHeight: '1.2',
              textAlign: 'left',
            },
            content: [
              {
                type: 'text',
                marks: [
                  {
                    type: 'bold',
                  },
                  {
                    type: 'textStyle',
                    attrs: {
                      fontFamily: 'Merriweather',
                      fontSize: '24px',
                      color: '#000',
                    },
                  },
                ],
                text: 'MARKET CONDITIONS',
              },
            ],
          },
          {
            type: 'title',
            attrs: {
              lineHeight: '1.2',
              textAlign: 'left',
            },
          },
          {
            type: 'title',
            attrs: {
              lineHeight: '1.2',
              textAlign: 'left',
            },
            content: [
              {
                type: 'text',
                marks: [
                  {
                    type: 'textStyle',
                    attrs: {
                      fontFamily: 'Raleway',
                      fontSize: '15px',
                      color: '#000',
                    },
                  },
                ],
                text: 'The market cannot be fully predicted, especially in ones that rapidly grow. Keeping up with the market requires consistent research',
              },
            ],
          },
          {
            type: 'title',
            attrs: {
              lineHeight: '1.2',
              textAlign: 'left',
            },
            content: [
              {
                type: 'text',
                marks: [
                  {
                    type: 'textStyle',
                    attrs: {
                      fontFamily: 'Raleway',
                      fontSize: '15px',
                      color: '#000',
                    },
                  },
                ],
                text: 'and development to ensure that changes in strategies or visions are dutifully implemented. Further, aspects such as inflation or the cost of capital would have to be forecasted with uncertainties.',
              },
            ],
          },
        ],
      },
    },
    fontsToLoad: ['Raleway', 'Merriweather'],
    thumbnail: {
      src: MarketConditionsThumbnail,
      altText: 'Market conditions',
    },
  },
  // Additional units
  {
    widthPx: 351,
    heightPx: 113.2,
    backgroundShapeData: {
      fillColor: ['#fff'],
      cornerRadius: 10,
      border: {
        color: '#DB654E',
        width: 5,
        style: BorderStyle.Solid,
      },
    },
    textWidgetData: {
      widthPx: 311,
      heightPx: 73.2,
      proseMirrorData: {
        type: 'doc',
        content: [
          {
            type: 'paragraph',
            attrs: {
              lineHeight: '1.2',
              textAlign: 'center',
            },
            content: [
              {
                type: 'text',
                marks: [
                  {
                    type: 'textStyle',
                    attrs: {
                      fontFamily: 'Libre Franklin',
                      fontSize: '15px',
                      color: '#2B2B35',
                    },
                  },
                ],
                text: 'Additional units sold since last month',
              },
            ],
          },
          {
            type: 'title',
            attrs: {
              lineHeight: '1.2',
              textAlign: 'center',
            },
            content: [
              {
                type: 'text',
                marks: [
                  {
                    type: 'bold',
                  },
                  {
                    type: 'textStyle',
                    attrs: {
                      fontFamily: 'Libre Franklin',
                      fontSize: '45px',
                      color: '#DB654E',
                    },
                  },
                ],
                text: '1,269,000',
              },
            ],
          },
        ],
      },
    },
    fontsToLoad: ['Libre Franklin'],
    thumbnail: {
      src: AdditionalUnitsThumbnail,
      altText: 'Additional units',
    },
  },
  // Quote
  {
    widthPx: 282,
    heightPx: 116.78,
    backgroundShapeData: {
      fillColor: ['#fff'],
      cornerRadius: 10,
      border: {
        color: '#24C0CF',
        width: 4,
        style: BorderStyle.Dashed,
      },
    },
    textWidgetData: {
      widthPx: 242,
      heightPx: 76.78,
      proseMirrorData: {
        type: 'doc',
        content: [
          {
            type: 'paragraph',
            attrs: {
              textAlign: 'center',
              lineHeight: '1.2',
            },
            content: [
              {
                text: '“I was treated with respect and the support team was quick to solve my problem. Definitely will recommend this service!”',
                type: 'text',
                marks: [
                  {
                    type: 'textStyle',
                    attrs: {
                      color: '#2B2B35',
                      fontSize: '15px',
                      fontFamily: 'Inter',
                    },
                  },
                ],
              },
            ],
          },
        ],
      },
    },
    thumbnail: {
      src: QuoteThumbnail,
      altText: 'Quote',
    },
  },
];
