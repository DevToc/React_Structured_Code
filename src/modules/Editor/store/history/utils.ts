export const isEmpty = (value: any): boolean => {
  return (
    value === null ||
    value === undefined ||
    (typeof value === 'object' && !Object.values(value).some((v) => v !== null && typeof v !== 'undefined')) ||
    (typeof value === 'string' && value.length === 0)
  );
};

export const isNotEmpty = (value: any) => !isEmpty(value);
