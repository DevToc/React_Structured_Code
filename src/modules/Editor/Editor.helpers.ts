import { IgnoreEventElement, IgnoreEventRole } from './Editor.config';

export const shouldIgnoreEventTarget = (target: HTMLElement): boolean => {
  if (!target) return false;
  if (!target.getAttribute) return false;
  if (target.getAttribute('data-allow-global-event')) return false;

  const tagName = target?.tagName as IgnoreEventElement;
  if (tagName && IgnoreEventElement[tagName]) return true;

  const role = target.getAttribute('role') as IgnoreEventRole;
  if (role && IgnoreEventRole[role]) return true;

  return false;
};
