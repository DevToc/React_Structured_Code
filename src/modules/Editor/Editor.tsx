import { ReactElement, Suspense } from 'react';
import { Provider } from 'react-redux';
import { Flex, Spinner } from '@chakra-ui/react';

import { useDynamicImport } from 'hooks/useDynamicImport';
import { useViewport } from 'hooks/useViewport';

import { Toolbar } from './components/Toolbar';
import { IntercomProvider } from '../../libs/third-party/Intercom/intercomProvider';

import { CheckerSettingsProvider } from './components/AccessibilityManager/AccessibilityChecker/checker.provider';
import { AccessibilityMenu } from './components/AccessibilityManager/AccessibilityManager';
import { BoundingBoxProvider } from './components/BoundingBox/useBoundingBox';
import { Clipboard } from './components/Clipboard';
import { FlowModeProvider } from './components/FlowCore/FlowModeProvider';
import { Focus, FocusProvider } from './components/Focus';
import { Head } from './components/Head';
import { KeyboardShortcut } from './components/KeyboardShortcut';
import { SideMenuSettings } from './components/SideMenuSettings/SideMenuSettings';
import { SkipLinks } from './components/SkipLinks';
import { InContextToolbar } from './components/InContextToolbar';
import { ResponsiveComponentsMap } from './components.lazy';

import { store } from './store';
import { PersistManagerProvider } from './store/persistManager';
import { HistoryManagerProvider } from './store/history/historyManager';

import { InfographLoader } from '../InfographLoader';

const Editor = (): ReactElement => {
  const { isMobile, viewportHeight } = useViewport();

  const ResponsiveComponent = (name: string): ReactElement => {
    const Component = useDynamicImport(name, ResponsiveComponentsMap);
    return (
      <Suspense fallback={<Spinner />}>
        <Component />
      </Suspense>
    );
  };

  const Navbar = (): ReactElement => ResponsiveComponent('navbar');
  const PageArea = (): ReactElement => ResponsiveComponent('pageArea');
  const WidgetMenu = (): ReactElement => ResponsiveComponent('widget');

  return (
    <IntercomProvider>
      <Provider store={store}>
        <Head />
        <SkipLinks />
        <BoundingBoxProvider>
          <HistoryManagerProvider>
            <FocusProvider>
              <CheckerSettingsProvider>
                <PersistManagerProvider>
                  <Navbar />
                </PersistManagerProvider>
                <AccessibilityMenu />
              </CheckerSettingsProvider>
              <InfographLoader>
                <Flex as='main' minH={viewportHeight}>
                  <WidgetMenu />
                  <FlowModeProvider>
                    <Flex flex='1' direction='column' position='relative'>
                      {!isMobile && <Toolbar />}
                      <PageArea />
                      {isMobile && <InContextToolbar />}
                    </Flex>
                  </FlowModeProvider>
                  <SideMenuSettings />
                </Flex>
                <KeyboardShortcut />
                <Clipboard />
                <Focus />
              </InfographLoader>
            </FocusProvider>
          </HistoryManagerProvider>
        </BoundingBoxProvider>
      </Provider>
    </IntercomProvider>
  );
};

export default Editor;
