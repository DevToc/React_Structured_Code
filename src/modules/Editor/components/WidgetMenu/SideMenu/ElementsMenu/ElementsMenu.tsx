import { useRef } from 'react';
import { GridItem, Grid, Flex } from '@chakra-ui/react';
import { loadFonts } from 'hooks/useFont';
import { addNewWidget, useAppSelector, useAppDispatch } from 'modules/Editor/store';
import { selectZoom } from 'modules/Editor/store/selectEditorSettings';
import { selectInfographWidthPx } from 'modules/Editor/store/infographSelector';
import { STYLED_RESPONSIVE_TEXT_WIDGETS } from './ElementsMenu.config';
import { generateDefaultData } from 'widgets/ResponsiveWidgets/ResponsiveTextWidget/ResponsiveTextWidget.helpers';
import { ElementThumbnail } from './ElementThumbnail';
import { calculateInitialTopPx, INITIAL_LEFTPX_RATIO } from 'utils/calculateInitialWidgetPos';

interface ElementsMenuProps {
  onComplete?: () => void;
}

export const ElementsMenu = ({ onComplete }: ElementsMenuProps) => {
  // font loading is async, so we need to prevent multiple widgets being added at once if the user clicks quickly while the fonts are loading
  const widgetsBeingAddedMap = useRef<{ [index: number]: boolean }>({});
  const dispatch = useAppDispatch();

  const zoom = useAppSelector(selectZoom);
  const infographWidthPx = useAppSelector(selectInfographWidthPx);

  const doAddWidget = async (index: number) => {
    if (widgetsBeingAddedMap.current[index]) return;
    widgetsBeingAddedMap.current[index] = true;

    const styleData = STYLED_RESPONSIVE_TEXT_WIDGETS[index];
    const { widthPx, heightPx, fontsToLoad } = styleData;
    const initialTopPx = calculateInitialTopPx(zoom);
    const initialLeftPx = INITIAL_LEFTPX_RATIO * infographWidthPx;

    if (fontsToLoad && fontsToLoad.length > 0) {
      await loadFonts(fontsToLoad);
    }

    dispatch(
      addNewWidget(
        generateDefaultData({
          shapeDataOverride: { widthPx, heightPx, ...styleData.backgroundShapeData },
          textDataOverride: styleData.textWidgetData,
          parentDataOverride: { widthPx, heightPx, topPx: initialTopPx, leftPx: initialLeftPx },
        }),
      ),
    );

    widgetsBeingAddedMap.current[index] = false;

    if (typeof onComplete === 'function') onComplete();
  };

  return (
    <Flex direction={'column'} p={4}>
      <Grid templateColumns={'repeat(2, 1fr)'} columnGap={4} rowGap={4}>
        {STYLED_RESPONSIVE_TEXT_WIDGETS.map((_, index) => (
          <GridItem key={index} h={120}>
            <ElementThumbnail onClick={doAddWidget} index={index} />
          </GridItem>
        ))}
      </Grid>
    </Flex>
  );
};
