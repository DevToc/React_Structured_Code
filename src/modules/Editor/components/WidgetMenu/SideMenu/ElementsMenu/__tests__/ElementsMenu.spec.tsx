import { screen, fireEvent } from '@testing-library/react';
import { renderWithRedux } from 'utils/test-utils.test';
import { ElementsMenu } from '../ElementsMenu';
import { InfographLoader } from 'modules/InfographLoader';
import Editor from 'modules/Editor';
import { EMPTY_INFOGRAPH } from 'utils/loadSampleInfograph';

describe('ElementsMenu', () => {
  const disableWarnings = () => {
    // TODO figure out why this error is thrown for tests but not in browser
    // Suppress error 'Received NaN for data-rotation attribute'
    jest.spyOn(console, 'error').mockImplementation((message) => {
      if (message.includes('Received NaN')) return;
      else console.error(message);
    });
  };

  it('should render', () => {
    const { asFragment } = renderWithRedux(<ElementsMenu />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should block adding same widget if loading widget', async () => {
    disableWarnings();
    renderWithRedux(
      <InfographLoader infographState={EMPTY_INFOGRAPH}>
        <Editor />
      </InfographLoader>,
    );

    const elementMenu = screen.getByTestId('Elements');
    fireEvent.click(elementMenu);

    const addElementButton = screen.getByTestId('elementsmenu-0-btn');
    fireEvent.click(addElementButton);
    fireEvent.click(addElementButton);

    const allWidgetsAfter = await screen.findAllByTestId(/widgetbase/);

    // the element is a responsive widget and contains 3 widgets
    // only one element should be added
    expect(allWidgetsAfter.length).toBe(3);
  });
});
