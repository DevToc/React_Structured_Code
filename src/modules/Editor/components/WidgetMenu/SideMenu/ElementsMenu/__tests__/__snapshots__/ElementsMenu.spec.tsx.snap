// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElementsMenu should render 1`] = `
<DocumentFragment>
  <div
    class="css-1psu3b8"
  >
    <div
      class="css-1rxdf5q"
    >
      <div
        class="css-6gxhc5"
      >
        <button
          class="css-grwglm"
          data-testid="elementsmenu-0-btn"
        >
          <img
            alt="Set your goals thumbnail"
            class="chakra-image css-kr4z"
            src="set_your_goals.png"
          />
        </button>
      </div>
      <div
        class="css-6gxhc5"
      >
        <button
          class="css-grwglm"
          data-testid="elementsmenu-1-btn"
        >
          <img
            alt="1.3 Million users"
            class="chakra-image css-kr4z"
            src="monthly_active_users.png"
          />
        </button>
      </div>
      <div
        class="css-6gxhc5"
      >
        <button
          class="css-grwglm"
          data-testid="elementsmenu-2-btn"
        >
          <img
            alt="Design button"
            class="chakra-image css-kr4z"
            src="design.png"
          />
        </button>
      </div>
      <div
        class="css-6gxhc5"
      >
        <button
          class="css-grwglm"
          data-testid="elementsmenu-3-btn"
        >
          <img
            alt="25.5% of people"
            class="chakra-image css-kr4z"
            src="referral_traffic.png"
          />
        </button>
      </div>
      <div
        class="css-6gxhc5"
      >
        <button
          class="css-grwglm"
          data-testid="elementsmenu-4-btn"
        >
          <img
            alt="Analysis"
            class="chakra-image css-kr4z"
            src="analysis.png"
          />
        </button>
      </div>
      <div
        class="css-6gxhc5"
      >
        <button
          class="css-grwglm"
          data-testid="elementsmenu-5-btn"
        >
          <img
            alt="Market conditions"
            class="chakra-image css-kr4z"
            src="market_conditions.png"
          />
        </button>
      </div>
      <div
        class="css-6gxhc5"
      >
        <button
          class="css-grwglm"
          data-testid="elementsmenu-6-btn"
        >
          <img
            alt="Additional units"
            class="chakra-image css-kr4z"
            src="additional_units.png"
          />
        </button>
      </div>
      <div
        class="css-6gxhc5"
      >
        <button
          class="css-grwglm"
          data-testid="elementsmenu-7-btn"
        >
          <img
            alt="Quote"
            class="chakra-image css-kr4z"
            src="quote.png"
          />
        </button>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
